vit_pytorch/masked_prediction_loss.py

```diff
@@ -0,0 +1,55 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class MaskedPredictionLoss(nn.Module):
+    def __init__(self,
+                 patch_size,
+                 img_size,
+                 device="cpu"):
+        super(MaskedPredictionLoss, self).__init__()
+        self.patch_size = patch_size
+        self.num_patch_axis = img_size // patch_size
+
+    def _transform_targets(self, targets, masked_patches):
+        masked_patches_dims = []
+        for masked_patch in masked_patches:
+            height_offset = self.patch_size * ((masked_patch) // self.num_patch_axis)
+            width_offset = self.patch_size * ((masked_patch) % self.num_patch_axis)
+            masked_patches_dims.append([height_offset, width_offset])
+
+        target_patches = []
+        for target in range(targets.shape[0]):
+            for masked_patch in masked_patches_dims:
+                height_offset, width_offset = masked_patch
+                extracted_patch = targets[target, :,
+                                        height_offset:height_offset + self.patch_size,
+                                        width_offset:width_offset + self.patch_size]
+                target_patches.append(extracted_patch)
+
+        target_patches_tensor = torch.stack(target_patches)
```

Comment:
all the above can probably be done with einops and then indexing

```python
from einops import rearrange

target_patches_tensor = rearrange(targets, 'b c (h p1) (w p2) -> b (h w) c h w', p1 = self.patch_size, p2 = self.patch_size)
target_patches_tensor = target_patches_tensor[:, masked_patches]
```

or something like that

Reply:
Ok cool, I haven't seen einops before but I'll look into that. Thanks for the suggestions!